### DISCUSSION TITLE
How to handle device disconnections?

### DISCUSSION THREAD

**Discussion by mihai-cristian3** [PhilipsHue/flutter_reactive_ble#563] [Ideas]
Hi community,

First of all, thank you so much for this incredible library. I've used many Flutter BLE libraries and I really think this one is THE ONE.

But I come here with an important question: how can we handle device disconnections? I mean, situations like lose of power, of signal (because of limited range), etc. I'm trying to deal with this but as for today, I still couldn't get it right. Does anybody have any clue?

Thank you all in advance.

**Comment by mad-mikey** [comment]
Hi,
I'm searching a way to handle this problem too.
An idea it could be using a sort of "watchdog", in which you continue to read one characteristic, for example every second, and if after some time you don't receive any response you could assume that the device is going out of range or it turns off. As a result you could execute a disconnection, without leaving pending connections.

I'm not sure it could help you this solution, I hope so.

**Comment by akospwc** [comment]
We are using https://pub.dev/documentation/reactive_ble_platform_interface/latest/reactive_ble_platform_interface/ReactiveBlePlatform/connectionUpdateStream.html to listen to connection state changes. No complaints.

**Reply by farr64** [reply]
It's nice to know about this package. The description is great:

- - -
A common platform interface for the [reactive ble](https://github.com/PhilipsHue/flutter_reactive_ble/) plugin. This package ensures every platform specific implementation uses the same interface.
- - -

Any specific usage examples anywhere?

The documentation at https://pub.dev/packages/reactive_ble_platform_interface is terse and there is no `Example`.

Thanks.

**Reply by akospwc** [reply]
It works exactly like the connection-specific status listener documented here: https://pub.dev/packages/flutter_reactive_ble#establishing-connection . The difference is that it delivers connection state updates for all active BLE connections, which is convenient.

**Reply by farr64** [reply]
Great. Thanks, @akospwc .

**Comment by mihai-cristian3** [comment]
I'm using the example app from flutter and I'm running into an interesting issue connecting to an ESP32.

When I disconnect manually from the device, everything goes well. I am able to reconnect afterwards and so on. But, when something happens, let's say I turn off the ESP32 or my phone loses the connection, I get these error in Android Studio.

> D/BluetoothGatt(26971): cancelOpen() - device: 84:CC:A8:5F:7E:2A
D/BluetoothGatt(26971): onClientConnectionState() - status=0 clientIf=9 device=84:CC:A8:5F:7E:2A
D/BluetoothGatt(26971): setCharacteristicNotification() - uuid: 3b70990b-ec67-47ef-b2ed-c985eef77ad4 enable: false
D/BluetoothGatt(26971): setCharacteristicNotification() - uuid: ff683b56-3159-4e51-a494-693ede3dc59b enable: false
D/BluetoothGatt(26971): setCharacteristicNotification() - uuid: 5687cc6f-b876-4b82-800f-f87b82746cb7 enable: false
W/System.err(26971): io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | com.polidea.rxandroidble2.exceptions.BleDisconnectedException: Disconnected from MAC='XX:XX:XX:XX:XX:XX' with status 0 (GATT_SUCCESS)
W/System.err(26971): 	at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.dispose(ObservableFlatMap.java:313)
W/System.err(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableReplay$ReplayObserver.dispose(ObservableReplay.java:271)
W/System.err(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableRefCount.timeout(ObservableRefCount.java:156)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableRefCount.cancel(ObservableRefCount.java:103)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableRefCount$RefCountObserver.dispose(ObservableRefCount.java:252)
W/System.err(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
W/System.err(26971): 	at io.reactivex.internal.operators.mixed.CompletableAndThenObservable$AndThenObservableObserver.dispose(CompletableAndThenObservable.java:86)
W/System.err(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
W/System.err(26971): 	at io.reactivex.internal.operators.mixed.SingleFlatMapObservable$FlatMapObserver.dispose(SingleFlatMapObservable.java:84)
W/System.err(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
W/System.err(26971): 	at io.reactivex.internal.operators.mixed.SingleFlatMapObservable$FlatMapObserver.dispose(SingleFlatMapObservable.java:84)
W/System.err(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$InnerObserver.dispose(ObservableFlatMap.java:588)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.disposeAll(ObservableFlatMap.java:510)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.dispose(ObservableFlatMap.java:310)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.disposeAll(ObservableFlatMap.java:504)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.dispose(ObservableFlatMap.java:310)
W/System.err(26971): 	at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.dispose(ObservableObserveOn.java:146)
W/System.err(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
W/System.err(26971): 	at io.reactivex.internal.observers.LambdaObserver.dispose(LambdaObserver.java:102)
W/System.err(26971): 	at com.signify.hue.flutterreactiveble.channelhandlers.CharNotificationHandler.unsubscribeFromAllNotifications(CharNotificationHandler.kt:59)
W/System.err(26971): 	at com.signify.hue.flutterreactiveble.channelhandlers.CharNotificationHandler.onCancel(CharNotificationHandler.kt:28)
W/System.err(26971): 	at io.flutter.plugin.common.EventChannel$IncomingStreamRequestHandler.onCancel(EventChannel.java:231)
W/System.err(26971): 	at io.flutter.plugin.common.EventChannel$IncomingStreamRequestHandler.onMessage(EventChannel.java:199)
W/System.err(26971): 	at io.flutter.embedding.engine.dart.DartMessenger.invokeHandler(DartMessenger.java:296)
W/System.err(26971): 	at io.flutter.embedding.engine.dart.DartMessenger.lambda$dispatchMessageToQueue$0$DartMessenger(DartMessenger.java:320)
W/System.err(26971): 	at io.flutter.embedding.engine.dart.-$$Lambda$DartMessenger$TsixYUB5E6FpKhMtCSQVHKE89gQ.run(Unknown Source:12)
W/System.err(26971): 	at android.os.Handler.handleCallback(Handler.java:938)
W/System.err(26971): 	at android.os.Handler.dispatchMessage(Handler.java:99)
D/BluetoothGatt(26971): close()
D/BluetoothGatt(26971): unregisterApp() - mClientIf=9
W/System.err(26971): 	at android.os.Looper.loopOnce(Looper.java:201)
W/System.err(26971): 	at android.os.Looper.loop(Looper.java:288)
W/System.err(26971): 	at android.app.ActivityThread.main(ActivityThread.java:7870)
W/System.err(26971): 	at java.lang.reflect.Method.invoke(Native Method)
W/System.err(26971): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
W/System.err(26971): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
W/System.err(26971): Caused by: com.polidea.rxandroidble2.exceptions.BleDisconnectedException: Disconnected from MAC='XX:XX:XX:XX:XX:XX' with status 0 (GATT_SUCCESS)
W/System.err(26971): 	at com.polidea.rxandroidble2.internal.connection.RxBleGattCallback$2.onConnectionStateChange(RxBleGattCallback.java:81)
W/System.err(26971): 	at android.bluetooth.BluetoothGatt$1$4.run(BluetoothGatt.java:281)
W/System.err(26971): 	at android.bluetooth.BluetoothGatt.runOrQueueCallback(BluetoothGatt.java:823)
W/System.err(26971): 	at android.bluetooth.BluetoothGatt.access$200(BluetoothGatt.java:47)
W/System.err(26971): 	at android.bluetooth.BluetoothGatt$1.onClientConnectionState(BluetoothGatt.java:276)
W/System.err(26971): 	at android.bluetooth.IBluetoothGattCallback$Stub.onTransact(IBluetoothGattCallback.java:201)
W/System.err(26971): 	at android.os.Binder.execTransactInternal(Binder.java:1184)
W/System.err(26971): 	at android.os.Binder.execTransact(Binder.java:1143)
E/AndroidRuntime(26971): FATAL EXCEPTION: main
E/AndroidRuntime(26971): Process: com.example.reactive_scoreboard_ble, PID: 26971
E/AndroidRuntime(26971): io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | com.polidea.rxandroidble2.exceptions.BleDisconnectedException: Disconnected from MAC='XX:XX:XX:XX:XX:XX' with status 0 (GATT_SUCCESS)
E/AndroidRuntime(26971): 	at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.dispose(ObservableFlatMap.java:313)
E/AndroidRuntime(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableReplay$ReplayObserver.dispose(ObservableReplay.java:271)
E/AndroidRuntime(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableRefCount.timeout(ObservableRefCount.java:156)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableRefCount.cancel(ObservableRefCount.java:103)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableRefCount$RefCountObserver.dispose(ObservableRefCount.java:252)
E/AndroidRuntime(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.mixed.CompletableAndThenObservable$AndThenObservableObserver.dispose(CompletableAndThenObservable.java:86)
E/AndroidRuntime(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.mixed.SingleFlatMapObservable$FlatMapObserver.dispose(SingleFlatMapObservable.java:84)
E/AndroidRuntime(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.mixed.SingleFlatMapObservable$FlatMapObserver.dispose(SingleFlatMapObservable.java:84)
E/AndroidRuntime(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$InnerObserver.dispose(ObservableFlatMap.java:588)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.disposeAll(ObservableFlatMap.java:510)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.dispose(ObservableFlatMap.java:310)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.disposeAll(ObservableFlatMap.java:504)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.dispose(ObservableFlatMap.java:310)
E/AndroidRuntime(26971): 	at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.dispose(ObservableObserveOn.java:146)
E/AndroidRuntime(26971): 	at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
E/AndroidRuntime(26971): 	at io.reactivex.internal.observers.LambdaObserver.dispose(LambdaObserver.java:102)
E/AndroidRuntime(26971): 	at com.signify.hue.flutterreactiveble.channelhandlers.CharNotificationHandler.unsubscribeFromAllNotifications(CharNotificationHandler.kt:59)
E/AndroidRuntime(26971): 	at com.signify.hue.flutterreactiveble.channelhandlers.CharNotificationHandler.onCancel(CharNotificationHandler.kt:28)
E/AndroidRuntime(26971): 	at io.flutter.plugin.common.EventChannel$IncomingStreamRequestHandler.onCancel(EventChannel.java:231)
E/AndroidRuntime(26971): 	at io.flutter.plugin.common.EventChannel$IncomingStreamRequestHandler.onMessage(EventChannel.java:199)
E/AndroidRuntime(26971): 	at io.flutter.embedding.engine.dart.DartMessenger.invokeHandler(DartMessenger.java:296)
E/AndroidRuntime(26971): 	at io.flutter.embedding.engine.dart.DartMessenger.lambda$dispatchMessageToQueue$0$DartMessenger(DartMessenger.java:320)
E/AndroidRuntime(26971): 	at io.flutter.embedding.engine.dart.-$$Lambda$DartMessenger$TsixYUB5E6FpKhMtCSQVHKE89gQ.run(Unknown Source:12)
E/AndroidRuntime(26971): 	at android.os.Handler.handleCallback(Handler.java:938)
E/AndroidRuntime(26971): 	at android.os.Handler.dispatchMessage(Handler.java:99)
E/AndroidRuntime(26971): 	at android.os.Looper.loopOnce(Looper.java:201)
E/AndroidRuntime(26971): 	at android.os.Looper.loop(Looper.java:288)
E/AndroidRuntime(26971): 	at android.app.ActivityThread.main(ActivityThread.java:7870)
E/AndroidRuntime(26971): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(26971): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
E/AndroidRuntime(26971): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
E/AndroidRuntime(26971): Caused by: com.polidea.rxandroidble2.exceptions.BleDisconnectedException: Disconnected from MAC='XX:XX:XX:XX:XX:XX' with status 0 (GATT_SUCCESS)
E/AndroidRuntime(26971): 	at com.polidea.rxandroidble2.internal.connection.RxBleGattCallback$2.onConnectionStateChange(RxBleGattCallback.java:81)
E/AndroidRuntime(26971): 	at android.bluetooth.BluetoothGatt$1$4.run(BluetoothGatt.java:281)
E/AndroidRuntime(26971): 	at android.bluetooth.BluetoothGatt.runOrQueueCallback(BluetoothGatt.java:823)
E/AndroidRuntime(26971): 	at android.bluetooth.BluetoothGatt.access$200(BluetoothGatt.java:47)
E/AndroidRuntime(26971): 	at android.bluetooth.BluetoothGatt$1.onClientConnectionState(BluetoothGatt.java:276)
E/AndroidRuntime(26971): 	at android.bluetooth.IBluetoothGattCallback$Stub.onTransact(IBluetoothGattCallback.java:201)
E/AndroidRuntime(26971): 	at android.os.Binder.execTransactInternal(Binder.java:1184)
E/AndroidRuntime(26971): 	at android.os.Binder.execTransact(Binder.java:1143)
I/Process (26971): Sending signal. PID: 26971 SIG: 9
Lost connection to device.


The app stops running. I don't know how to handle this, what is causing it or how I can approach it.

**Reply by akospwc** [reply]
Just a wild guess: are you experiencing the issue described in the package's FAQ ("How to handle the BLE undeliverable exception")?

https://pub.dev/packages/flutter_reactive_ble#faq
